File: scala/src/main/scala/kip/util/QHull.scala
package kip.util

import java.io._

object QHull {
  
  case class Vertices2d(a: Array[Double]) {
    val n: Int = a.size/2
    def x(i: Int): Double = a(2*i+0)
    def y(i: Int): Double = a(2*i+1)
  }
  
  def delaunay2d(vertices: Vertices2d): Array[Int] = {
    val p = Runtime.getRuntime().exec("qdelaunay QJ i Fx")
    val pw = new PrintWriter(p.getOutputStream())
    val br = new BufferedReader(new InputStreamReader(p.getInputStream()));
    
    // write vertices to qhull
    pw.println(2) // dimension
    pw.println(vertices.n)
    for (i <- 0 until vertices.n) {
      pw.println(vertices.x(i)+" "+vertices.y(i))
    }
    pw.close()
    
    // read triangle output from "i" option
    val ntris = br.readLine().toInt
    val tris = new Array[Int](3*ntris)
    for (i <- 0 until ntris) {
      val Array(v1, v2, v3) = br.readLine().split("\\s+")
      tris(3*i+0) = v1.toInt
      tris(3*i+1) = v2.toInt
      tris(3*i+2) = v3.toInt
    }
    
    tris
  }
  

  def test() = {
    val a = Array[Double](0,0, 0,1, 1,1, 1,0, 0.6, 0.7)
    delaunay2d(Vertices2d(a))
  }
}

File: src/main/java/kip/javasim/geometry/QHull.java
package kip.javasim.geometry;

import java.io.*;

import scikit.util.Point;


public class QHull {
	private String path;
	private PrintWriter pw;
	private BufferedReader br;
	
	public QHull(String path) {
		this.path = path;
	}
	
	public Point[][] constructVoronoi2D(double[] state, int stride, int N0, int N1) {
		try {
			Process p = Runtime.getRuntime().exec(path + " v o");
			pw = new PrintWriter(p.getOutputStream());
			br = new BufferedReader(new InputStreamReader(p.getInputStream()));
		} catch (IOException e) {
			System.err.println("QHull not available at " + path);
		}

		pw.println(2); // dimension
		pw.println(N1-N0); // number of points
		for (int i = N0; i < N1; i++) {
			pw.println(state[(2*i+0)*stride] + " " + state[(2*i+1)*stride]); // x y coordinates
		}
		pw.close();
		
		try {
			br.readLine(); // dimension
			String[] line = br.readLine().trim().split("\\s+");
			int nvertices = Integer.valueOf(line[0]);
			int nregions = Integer.valueOf(line[1]);
			
			Point vertices[] = new Point[nvertices];
			for (int i = 0; i < nvertices; i++) {
				line = br.readLine().trim().split("\\s+");
				vertices[i] = new Point(Double.valueOf(line[0]), Double.valueOf(line[1]));
			}
			vertices[0] = null; // this is the vertex at "infinity"
			
			Point regions[][] = new Point[nregions][];
			for (int i = 0; i < nregions; i++) {
				line = br.readLine().trim().split("\\s+");
				int npts = line.length-1;
				regions[i] = new Point[npts];
				for (int j = 0; j < npts; j++) {
					regions[i][j] = vertices[Integer.valueOf(line[j+1])];
				}
			}
			return regions;
		} catch (IOException e) {
			System.err.println("Invalid QHull output.");
		}
		return null;
	}
}
